### DISCUSSION TITLE
How to implement swap of buffers?

### DISCUSSION THREAD

**Discussion by Anshi701** [vsg-dev/VulkanSceneGraph#1054] [Q&A] [answered]
What im trying to do:
 - Have a big arrays of vertex data for any amount tiles of my surface.
 - Update part of that data or add new tile data from another threads aside from render thread. That operation must not affect the render.
 - Change buffer i need to use for rendering (in Update stage i suppose).

Please, give me a hint how to do it properly in VSG-style way.

**Comment by robertosfield** [comment]
Personally I'd use vertex array data per tile, but have those vertex arrays use shared vk/vsg::Buffer.  The VSG is already capable of doing this, much of it transparently.

Doing this will allow you to create/update vertex array data per tile without worry about coordinating data structures across tiles, and if you do end up need to use more vsg::Buffer then these can be used as well.  

The VSG already has a vsg::PagedLOD and vsg::DatabasePager that support database pager of whole earth databases, as there is Tim's work on vsgCs which adds 3D Tiles support.  These both use vertex arrays/textures per tile and let the underlying support for sharing vsg::Buffer between multiple vertex arrays/uniforms - vsg::BufferInfo couples the CPU array data with the vk/vsg::Buffer and the location that that vsg::Buffer, and vsg::MemoryBufferPools is used to set up the BufferInfo to share vk/vsg::Buffer.

--

Is there a particular terrain algorithm you are trying to implement?

**Reply by Anshi701** [reply]
I'm implementing my own database format, so i can not use those tools. 
Question is how to change data of vsg array from another thread without blocking render or updating data before operation ends.
As for using arrays per vertex, i will try to do that way, thank you for hint!

**Reply by robertosfield** [reply, marked as answer]
Dynamic data in the VSG is handled via the vsg::TransferTask, this multi-buffers the data and handles all the transfer and synchronization for you.  You just need to set the vsg::Data::properties.dataVariance to  DYNAMIC_DATA and call data->dirty() each time the data is updated.  Have a look at the vsgdynamic* examples in vsgExamples:

https://github.com/vsg-dev/vsgExamples/tree/master/examples/state

I have written up this topic here on this forum and the old vsg-users googlegroup so have a look through the archives for more discussion on this topic.

My plan is to keep refining vsg::DataTransfer to handle parts of the compile traversal work, and handle one of copying of data rather than relying the checking the data's modification count that's updated by Data:dirty().

--

W.r.t you own database format, what is you motivation here? Do you have any details of what you are trying to achieve?  It's hard to know what advice to give you without knowing what you want to achieve and why.

**Reply by Anshi701** [reply]
Oh, i see, so data is already multi-buffered, so i can just change it anytime. Thanks for your explanation and patience!
My motivation is simple - it's needed for my job, company develop it's own format for some special tasks.
